docs/architecture/adr-038-state-listening.md

```diff
@@ -588,60 +378,88 @@ func NewSimApp(
+	if err := pluginLoader.Inject(bApp, appCodec, keys); err != nil {
+		tmos.Exit(err.Error())
+    }
+
+	// start the plugin services, optionally use wg to synchronize shutdown using io.Closer
+	wg := new(sync.WaitGroup)
+	if err := pluginLoader.Start(wg); err != nil {
+		tmos.Exit(err.Error())
+    }
 
 	...
 
 	return app
 }
 ```
 
+
+#### Configuration
+
+The plugin system will be configured within an app's app.toml file.
+
+```toml
+[plugins]
+    on = false # turn the plugin system, as a whole, on or off
+    disabled = ["list", "of", "plugin", "names", "to", "disable"]
+    dir = "the directory to load non-preloaded plugins from; defaults to "
+```
+
+There will be three parameters for configuring the plugin system: `plugins.on`, `plugins.disabled` and `plugins.dir`.
+`plugins.on` is a bool that turns on or off the plugin system at large, `plugins.dir` directs the system to a directory
+to load plugins from, and `plugins.disabled` is a list of names for the plugins we want to disable (useful for disabling preloaded plugins).
+
+Configuration of a given plugin is ultimately specific to the plugin, but we will introduce some standards here:
+
+Plugin TOML configuration should be split into separate sub-tables for each kind of plugin (e.g. `plugins.streaming`).
+Within these sub-tables, the parameters for a specific plugin of that kind are included in another sub-table (e.g. `plugins.streaming.file`).
+
+e.g.
+
+```toml
+[plugins]
+    on = false # turn the plugin system, as a whole, on or off
+    disabled = ["list", "of", "plugin", "names", "to", "disable"]
+    dir = "the directory to load non-preloaded plugins from; defaults to "
+    [plugins.streaming] # a mapping of plugin-specific streaming service parameters, mapped to their plugin name
+        [plugins.streaming.file] # the specific parameters for the file streaming service plugin
+            keys = ["list", "of", "store", "keys", "we", "want", "to", "expose", "for", "this", "streaming", "service"]
+            writeDir = "path to the write directory"
+            prefix = "optional prefix to prepend to the generated file names"
+        [plugins.streaming.kafka]
+            ...
+    [plugins.modules]
+        ...
+```
+
+#### Encoding and decoding streams
+
+ADR-038 introduces the interfaces and types for streaming state changes out from KVStores, associating this
+data with their related ABCI requests and responses, and registering a service for consuming this data and streaming it to some destination in a final format.
+Instead of prescribing a final data format in this ADR, it is left to a specific plugin implementation to define and document this format.
+We take this approach because flexibility in the final format is necessary to support a wide range of streaming service plugins. For example,
+the data format for a streaming service that writes the data out to a set of files will differ from the data format that is written to a Kafka topic.
+
```

Comment:
I think we should unify the data format for all external services, especially message queues. Essentially client app should easily connect to RabbitMQ or Redis and expect the same data format.

Reply:
The API already provides unified data by passing along the marshaled `[]byte` data. Adding a constraint like this might present some difficulties. The API needs be flexible to allow authors to choose the format that they want to store data on external systems. I don't see how it can be enforced given that plugin authors have control on how they `process`, `transform`, and `store` the data externally. They may only be interested in specific data stored in a `KVStore`, which can be `JSON`, `Protbuf`, or `String` values.